### PR TITLE
Add recipe for r-changepoints

### PR DESCRIPTION
**Pull request by mlondschien**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- ~~If static libraries are linked in, the license of the static library is packaged.~~
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there (me, not @conda-forge/r)
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/r-changepoints/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+source:
+  url:
+    - {{ cran_mirror }}/src/contrib/changepoints_{{ version }}.tar.gz
+    - {{ cran_mirror }}/src/contrib/Archive/changepoints/changepoints_{{ version }}.tar.gz
+  sha256: 73b7e389d8d5590d91ecb369f57b80f605b3c9f231da04be54a650b9b9c593f1
+
+build:
+  merge_build_host: true  # [win]
+  number: 0
+  noarch: generic
```
Comment (dbast) [inline recipes/r-changepoints/meta.yaml:18]:
```suggestion
```

File: recipes/r-changepoints/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  rpaths:
+    - lib/R/lib/
+    - lib/
+
+requirements:
+  build:
+    - {{ posix }}zip               # [win]
```
Comment (dbast) [inline recipes/r-changepoints/meta.yaml:25]:
```suggestion
    - libblas
    - {{ posix }}zip               # [win]
```

File: recipes/r-changepoints/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    - {{ compiler('m2w64_c') }}        # [win]
+    - {{ posix }}filesystem        # [win]
+    - {{ posix }}make
+    - {{ posix }}sed               # [win]
+    - {{ posix }}coreutils         # [win]
+    - {{ posix }}zip               # [win]
+    - cross-r-base {{ r_base }}    # [build_platform != target_platform]
+
+  host:
+    - r-base
```
Comment (dbast) [inline recipes/r-changepoints/meta.yaml:38]:
```suggestion
    - libblas
    - r-base
```

File: recipes/r-changepoints/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  host:
+    - libblas
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:44]:
```suggestion
    - r-rcpparmadillo
    - liblapack
```

File: recipes/r-changepoints/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  run:
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
+    - {{ native }}gcc-libs         # [win]
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:53]:
```suggestion
    - {{ native }}gcc-libs         # [win]
    - liblapack
```

File: recipes/r-changepoints/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    - liblapack
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:55]:
```suggestion
    - liblapack
    - libstdcxx-ng
```

File: recipes/r-changepoints/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,105 @@
+{% set name = "r-changepoints" %}
+{% set version = "1.1.0" %}
+{% set posix = 'm2-' if win else '' %}
+
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:4]:
```suggestion
{% set native = 'm2w64-' if win else '' %}
```

File: recipes/r-changepoints/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
+    - {{ native }}gcc-libs         # [win]
+    - liblapack
+    - libstdcxx-ng
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:56]:
```suggestion
    - libstdcxx-ng       # [linux]
    - libcxx                  # [osx]
```

File: recipes/r-changepoints/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  run:
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
+    - {{ native }}gcc-libs         # [win]
+    - liblapack
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:56]:
```suggestion
    - liblapack        # [not win]
```

File: recipes/r-changepoints/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  host:
+    - libblas
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
+    - liblapack
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:46]:
```suggestion
    - liblapack        # [not win]
```

File: recipes/r-changepoints/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    - {{ compiler('m2w64_c') }}        # [win]
+    - {{ posix }}filesystem        # [win]
+    - {{ posix }}make
+    - {{ posix }}sed               # [win]
+    - {{ posix }}coreutils         # [win]
+    - {{ posix }}zip               # [win]
+    - cross-r-base {{ r_base }}    # [build_platform != target_platform]
+
+  host:
+    - libblas
```
Comment (mlondschien) [inline recipes/r-changepoints/meta.yaml:38]:
```suggestion
    - libblas           # [not win]
```

**Comment by mlondschien** [comment]
Thanks, @dbast for your suggestions. The tests run through now, but the requirements are a bit convoluted. Could you have a look and let me know which ones are superfluous?

**Comment by mlondschien** [comment]
@conda-forge/r could you have a look?

**Comment by mlondschien** [comment]
@conda-forge/r  could you have a look?

**Comment by mlondschien** [comment]
@conda-forge/staged-recipes could you have a look?

File: recipes/r-changepoints/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    - {{ compiler('m2w64_c') }}        # [win]
+    - {{ posix }}filesystem        # [win]
+    - {{ posix }}make
+    - {{ posix }}sed               # [win]
+    - {{ posix }}coreutils         # [win]
+    - {{ posix }}zip               # [win]
+    - cross-r-base {{ r_base }}    # [build_platform != target_platform]
+
+  host:
+    - libblas           # [not win]
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:38]: Why not Windows? This deviates from typical recipes.

File: recipes/r-changepoints/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  run:
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
+    - {{ native }}gcc-libs         # [win]
+    - liblapack        # [not win]
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:56]: `liblapack` (like `libblas`) dependency should only need to be in `host:`.

File: recipes/r-changepoints/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+  merge_build_host: true  # [win]
+  number: 0
+  rpaths:
+    - lib/R/lib/
+    - lib/
+
+requirements:
+  build:
+    - {{ posix }}zip               # [win]
+    - r-rcpp
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:26]: Why are `r-rcpp` and `r-rcpparmadillo` under `build:`? These are usually only needed in `host:` and `run:`. They may be *conditionally* required while cross-compiling, but cross-compiling should be added only *after* feedstock creation.

File: recipes/r-changepoints/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
+    - {{ native }}gcc-libs         # [win]
+    - liblapack        # [not win]
+    - libstdcxx-ng       # [linux]
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:57]: Could you explain why the `libstdcxx-ng` and `libcxx` dependencies are needed at runtime? I've never seen this before.

File: recipes/r-changepoints/meta.yaml [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+        Wang, Yu and Willett (2020)
+    dependent dynamic nonparametric random dot product graphs
+        Padilla, Yu and Priebe (2019)
+    robust univariate mean against adversarial attacks
+        Li and Yu (2021)
+  license: GPL-3.0-only
+  license_family: GPL3
+  license_file:
+    - {{ environ["PREFIX"] }}/lib/R/share/licenses/GPL-3
+  doc_url: https://github.com/HaotianXu/changepoints
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:101]: If `doc_url` and `dev_url` are the same as `home`, then please omit them.

**Review by mfansler — CHANGES_REQUESTED** [review]
@mlondschien sorry for the delay. There are some odd things in this recipe that deviate from typically ones. I generated this recipe myself with `conda_r_skeleton_helper` (see [my `meta.yaml`](https://github.com/mfansler/staged-recipes/blob/15a7a1f540988f8b352b915bf6967a8efb5aa873/recipes/r-changepoints/meta.yaml)) and the only deviations I would have made were adding the `libblas` and `liblapack` to the `host:` requirements.

**Comment by mlondschien** [comment]
Thanks @mfansler. Could you have another look?

`libcxx` in host instead of run fails  [bd81a4a](https://github.com/conda-forge/staged-recipes/runs/10519540033).

File: recipes/r-changepoints/meta.yaml (hunk 15 above)
Comment (xhochy) [reply recipes/r-changepoints/meta.yaml:53]: This is an indicator that you are missing a `compiler('cxx')` build dependency.

File: recipes/r-changepoints/meta.yaml [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+  number: 0
+  rpaths:
+    - lib/R/lib/
+    - lib/
+
+requirements:
+  build:
+    - {{ compiler('c') }}          # [not win]
+    - {{ compiler('cxx') }}        # [not win]
+    - {{ compiler('m2w64_c') }}    # [win]
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:27]: Include `cxx` for win:

```suggestion
    - {{ compiler('m2w64_c') }}    # [win]
    - {{ compiler('m2w64_cxx') }}  # [win]
```

File: recipes/r-changepoints/meta.yaml [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    - cross-r-base {{ r_base }}    # [build_platform != target_platform]
+
+  host:
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:42]: `r-rcpp` is [an explicit dependency](https://github.com/HaotianXu/changepoints/blob/b96ed0c979ee155bb583d03a5023e9229801d8cf/DESCRIPTION#L23):

```suggestion
    - r-rcpp
    - r-rcpparmadillo
```

File: recipes/r-changepoints/meta.yaml [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    - liblapack        # [not win]
+    - libblas           # [not win]
+  run:
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:52]: `r-rcpp` is [an explicit dependency](https://github.com/HaotianXu/changepoints/blob/b96ed0c979ee155bb583d03a5023e9229801d8cf/DESCRIPTION#L23):

```suggestion
    - r-rcpp
    - r-rcpparmadillo
```

File: recipes/r-changepoints/meta.yaml [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+  host:
+    - r-base
+    - r-gglasso
+    - r-glmnet
+    - r-ks
+    - r-mass
+    - r-data.tree
+    - r-rcpparmadillo
+    - liblapack        # [not win]
+    - libblas           # [not win]
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:44]: Windows also [builds against BLAS/LAPACK](https://github.com/HaotianXu/changepoints/blob/b96ed0c979ee155bb583d03a5023e9229801d8cf/src/Makevars.win#L1):

```suggestion
    - liblapack
    - libblas
```

File: recipes/r-changepoints/meta.yaml [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+        Xu, Wang, Zhao, and Yu (2022)
+    high-dimensional vector autoregressive models
+        Wang, Yu, Rinaldo and Willett (2019)
+    high-dimensional self exciting point processes
+        Wang, Yu and Willett (2020)
+    dependent dynamic nonparametric random dot product graphs
+        Padilla, Yu and Priebe (2019)
+    robust univariate mean against adversarial attacks
+        Li and Yu (2021)
+  license: GPL-3.0-only
```
Comment (mfansler) [inline recipes/r-changepoints/meta.yaml:92]: [License is GPL (>= 3.0)](https://github.com/HaotianXu/changepoints/blob/b96ed0c979ee155bb583d03a5023e9229801d8cf/DESCRIPTION#L30):

```suggestion
  license: GPL-3.0-or-later
```

**Review by mfansler — CHANGES_REQUESTED** [review]
Still a few things outstanding.

**Review by mfansler — APPROVED** [review]

**Comment by mfansler** [comment]
@conda-forge/help-r I believe this is ready to merge

**Comment by mfansler** [comment]
@dbast do you mind revisiting this? I think it's now good to merge.

**Comment by mlondschien** [comment]
thanks @dbast @mfansler, and @xhochy

**Pull request merged**